src/hotspot/share/utilities/byteswap.hpp

```diff
@@ -0,0 +1,315 @@
+
+template <typename T, size_t N = sizeof(T)>
+struct ByteswapFallbackImpl;
+
+// We support 8-bit integer types to be compatible with C++23's std::byteswap.
+template <typename T>
+struct ByteswapFallbackImpl<T, 1> {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == 1);
+
+  ALWAYSINLINE T operator()(T x) const {
+    return x;
+  }
+};
+
+template <typename T>
+struct ByteswapFallbackImpl<T, 2> {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == 2);
+
+  ALWAYSINLINE uint16_t operator()(uint16_t x) const {
+    return (((x & UINT16_C(0x00ff)) << 8) | ((x & UINT16_C(0xff00)) >> 8));
+  }
+};
+
+template <typename T>
+struct ByteswapFallbackImpl<T, 4> {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == 4);
+
+  ALWAYSINLINE uint32_t operator()(uint32_t x) const {
+    return (((x & UINT32_C(0x000000ff)) << 24) | ((x & UINT32_C(0x0000ff00)) << 8) |
+            ((x & UINT32_C(0x00ff0000)) >> 8) | ((x & UINT32_C(0xff000000)) >> 24));
+  }
+};
+
+template <typename T>
+struct ByteswapFallbackImpl<T, 8> {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == 8);
+
+  ALWAYSINLINE uint64_t operator()(uint64_t x) const {
+    return (((x & UINT64_C(0x00000000000000ff)) << 56) | ((x & UINT64_C(0x000000000000ff00)) << 40) |
+            ((x & UINT64_C(0x0000000000ff0000)) << 24) | ((x & UINT64_C(0x00000000ff000000)) << 8) |
+            ((x & UINT64_C(0x000000ff00000000)) >> 8) | ((x & UINT64_C(0x0000ff0000000000)) >> 24) |
+            ((x & UINT64_C(0x00ff000000000000)) >> 40) | ((x & UINT64_C(0xff00000000000000)) >> 56));
+  }
+};
+
+/*****************************************************************************
+ * GCC and compatible (including Clang)
+ *****************************************************************************/
+#if defined(TARGET_COMPILER_gcc)
+
+#if defined(__clang__) || defined(ASSERT)
+
+// Unlike GCC, Clang is willing to inline the generic implementation of __builtin_bswap when
+// architecture support is unavailable in -O2. This ensures we avoid the function call to libgcc.
+// Clang is able to recognize the fallback implementation as byteswapping, but not on every
+// architecture unlike GCC. This suggests the optimization pass for GCC that recognizes byteswapping
+// is architecture agnostic, while for Clang it is not.
+
+// We support 8-bit integer types to be compatible with C++23's std::byteswap.
+template <typename T>
+struct ByteswapImpl<T, 1> final {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == 1);
+
+  ALWAYSINLINE T operator()(T x) const {
+    return x;
+  }
+};
+
+template <typename T>
+struct ByteswapImpl<T, 2> final {
+  STATIC_ASSERT(CanByteswapImpl<T>::value);
+  STATIC_ASSERT(sizeof(T) == sizeof(uint16_t));
+
+  ALWAYSINLINE T operator()(T x) const {
+    return static_cast<T>(__builtin_bswap16(static_cast<uint16_t>(x)));
```

Comment:
Hm. I believe we can. Originally I was worried about implicit integer promotion, for example if some compiler decided that uint32_t was unsigned int, but unsigned long was also the same size. It would promote unsigned long to unsigned long long causing issues. But we shouldn't have that here with the current approach.